File: TreeMachine.Pro_Cpp/TreeMachine.Pro.Tests/Sources/TreeMachine/Tests_03.cpp
#include "gtest/gtest.h"
#include "Tree.h"

namespace TreeMachine {
    using namespace std;

    TEST(Tests_04, Test_00) { // NOLINT
        auto *const root = new Root();
        auto *const a = new A();
        auto *const b = new B();

        {
            // AddChildren
            root->AddChildren(vector<Node *>{a, b}, nullptr);
            EXPECT_NE(root, nullptr);
            EXPECT_EQ(root->Tree(), nullptr);
            EXPECT_EQ(root->IsRoot(), true);
            // EXPECT_EQ(root->Root(), root);
            EXPECT_EQ(root->Parent(), nullptr);
            EXPECT_EQ(root->Activity(), Node::Activity_::Inactive);
            EXPECT_EQ(root->Children().size(), 2);
            for (const auto *const child : root->Children()) {
                EXPECT_EQ(child->Tree(), nullptr);
                EXPECT_EQ(child->IsRoot(), false);
                EXPECT_EQ(child->Root(), root);
                EXPECT_EQ(child->Parent(), root);
                EXPECT_EQ(child->Activity(), Node::Activity_::Inactive);
                EXPECT_EQ(child->Children().size(), 0);
            }
        }
        {
            // RemoveChildren
            root->RemoveChildren(nullptr, [](auto *const child, [[maybe_unused]] const auto arg) { delete child; });
            EXPECT_NE(root, nullptr);
            EXPECT_EQ(root->Tree(), nullptr);
            EXPECT_EQ(root->IsRoot(), true);
            // EXPECT_EQ(root->Root(), root);
            EXPECT_EQ(root->Parent(), nullptr);
            EXPECT_EQ(root->Activity(), Node::Activity_::Inactive);
            EXPECT_EQ(root->Children().size(), 0);
        }

        delete root;
    }

}

File: TreeMachine.Pro_CSharp/TreeMachine.Pro.Tests/System.TreeMachine/Tests_03.cs
namespace System.TreeMachine {
    using System;
    using System.Collections.Generic;
    using System.Text;
    using NUnit.Framework;
    using Assert = NUnit.Framework.Assert;

    public class Tests_03 {

        [Test]
        public void Test_00() {
            var root = new Root();
            var a = new A();
            var b = new B();
            {
                // root.AddChildren a, b
                root.AddChildren( [ a, b ], null );
                Assert.That( root.Children.Count, Is.EqualTo( 2 ) );
                foreach (var child in root.Children) {
                    Assert.That( child.Tree, Is.Null );
                    Assert.That( child.TreeRecursive, Is.Null );
                    Assert.That( child.Parent, Is.EqualTo( root ) );
                    Assert.That( child.Activity, Is.EqualTo( Node.Activity_.Inactive ) );
                    Assert.That( child.Children.Count, Is.EqualTo( 0 ) );
                }
            }
            {
                // root.RemoveChildren a, b
                root.RemoveChildren( null, null );
                Assert.That( root.Children.Count, Is.Zero );
            }
        }

    }
}
